store/localcachelayer/user_layer.go

```diff
@@ -127,9 +126,8 @@ func (s LocalCacheUserStore) GetProfileByIds(userIds []string, options *store.Us
 			return nil, model.NewAppError("SqlUserStore.GetProfileByIds", "store.sql_user.get_profiles.app_error", nil, err.Error(), http.StatusInternalServerError)
 		}
 
-		users = append(users, remainingUsers...)
-
 		for _, user := range remainingUsers {
+			users = append(users, user.DeepCopy())
```

Comment:
Do we need to deep copy for users just coming from the sql layer?

Reply:
We need a copy since we are also storing that data in the cache (line below). The idea is that we want the data pointed by the cache to be a copy of the data returned to avoid unwanted side effects.

Whether doing a deep copy or a shallow one is up for discussion. I prefer the former until we figure out a smarter way (immutables?). Trading some performance for data integrity seems like a good choice to me.